src/core/lru.h

```diff
@@ -0,0 +1,197 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#pragma once
+
+#include <absl/container/node_hash_map.h>
+
+#include <optional>
+
+#include "base/logging.h"
+#include "base/pmr/memory_resource.h"
+
+namespace dfly {
+
+enum class Position {
+  kHead,
+  kTail,
+};
+
+template <typename T, typename Hash> class Lru {
+  struct Node {
+    const T* data_ptr;
+
+    uint32_t prev;
+    uint32_t next;
+
+    Node() : prev(0), next(0) {
+    }
+  };
+
+ public:
+  explicit Lru(uint32_t capacity, PMR_NS::memory_resource* mr) : table_(mr), node_arr_(mr) {
+    CHECK_GT(capacity, 1u);
+    node_arr_.reserve(capacity);
+  }
+
+  std::optional<T> GetPrev(const T& data) const;
+  std::optional<T> GetTail() const;
+  std::optional<T> GetHead() const;
+
+  void Put(const T& data, Position position = Position::kHead);
+  bool Remove(const T& data);
+
+  size_t Size() const {
+    return table_.size();
```

Comment:
Consider adding `DCHECK_EQ(table_.size(), node_arr_.size())` - also potentially in other places